test/extensions/transport_sockets/tls/handshaker_test.cc

```diff
@@ -163,6 +165,83 @@ TEST_F(HandshakerTest, ErrorCbOnAbnormalOperation) {
   EXPECT_EQ(post_io_action, Network::PostIoAction::Close);
 }
 
+// Example HandshakerImpl demonstrating special-case behavior which necessitates
+// extra SSL_ERROR case handling. Here, we induce an SSL_ERROR_WANT_X509_LOOKUP,
+// check for it in the handshaker, faux-trigger the lookup, and then proceed as
+// normal.
+class HandshakerImplForTest : public Ssl::Handshaker {
+public:
+  HandshakerImplForTest(bssl::UniquePtr<SSL> ssl, std::function<void()> requested_cert_cb)
+      : ssl_(std::move(ssl)), requested_cert_cb_(requested_cert_cb) {}
+
+  Network::PostIoAction doHandshake(Ssl::SocketState& state,
+                                    Ssl::HandshakerCallbacks& callbacks) override {
+    ASSERT(state != Ssl::SocketState::HandshakeComplete && state != Ssl::SocketState::ShutdownSent);
+
+    int rc = SSL_do_handshake(ssl());
+    if (rc == 1) {
+      state = Ssl::SocketState::HandshakeComplete;
+      callbacks.OnSuccessCb(ssl());
+      return Network::PostIoAction::Close;
+    } else {
+      switch (SSL_get_error(ssl(), rc)) {
+      case SSL_ERROR_WANT_READ:
+      case SSL_ERROR_WANT_WRITE:
+        return Network::PostIoAction::KeepOpen;
+      case SSL_ERROR_WANT_X509_LOOKUP:
+        // Special case. Once this lookup is requested, we flip the bit and allow
+        // the handshake to proceed.
+        requested_cert_cb_();
+        return Network::PostIoAction::KeepOpen;
+      default:
+        callbacks.OnFailureCb();
+        return Network::PostIoAction::Close;
+      }
+    }
+  }
+
+  void setTransportSocketCallbacks(Network::TransportSocketCallbacks& callbacks) override {
+    transport_socket_callbacks_ = &callbacks;
+  }
+
+  SSL* ssl() override { return ssl_.get(); }
+
+private:
+  bssl::UniquePtr<SSL> ssl_;
+  std::function<void()> requested_cert_cb_;
+  Network::TransportSocketCallbacks* transport_socket_callbacks_{};
+};
+
+TEST_F(HandshakerTest, NormalOperationWithHandshakerImplForTest) {
+  bool cert_cb_ok = false;
+
+  ::testing::MockFunction<void()> requested_cert_cb;
+  EXPECT_CALL(requested_cert_cb, Call).WillOnce([&]() { cert_cb_ok = true; });
+
+  SSL_set_cert_cb(
```

Comment:
I'd move this to the constructor of the `HandshakerImplForTest`, as that would be an appropriate place for it if that were a real impl.

Reply:
Done.